networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRangeMessageHandler.java

```diff
@@ -0,0 +1,299 @@
+        LOG.trace("Stream closed while sending requested blobs sidecars", error);
+      } else {
+        LOG.error("Failed to process blob sidecars request", error);
+      }
+      callback.completeWithUnexpectedError(error);
+    }
+  }
+
+  private boolean verifyRequestedEpochIsInSupportedRange(
+      final UInt64 requestedSlot, final UInt64 minimumRequestEpoch) {
+    return spec.computeEpochAtSlot(requestedSlot).isGreaterThanOrEqualTo(minimumRequestEpoch);
+  }
+
+  private class RequestState {
+
+    private final AtomicInteger sentBlobSidecars = new AtomicInteger(0);
+    private final ResponseCallback<BlobSidecar> callback;
+    private final Bytes32 headBlockRoot;
+    private final AtomicReference<UInt64> currentSlot;
+    private final AtomicReference<UInt64> currentIndex;
+    private final UInt64 maxSlot;
+
+    private Optional<SignedBeaconBlock> maybeCurrentBlock = Optional.empty();
+
+    RequestState(
+        final ResponseCallback<BlobSidecar> callback,
+        final Bytes32 headBlockRoot,
+        final UInt64 currentSlot,
+        final UInt64 maxSlot) {
+      this.callback = callback;
+      this.headBlockRoot = headBlockRoot;
+      this.currentSlot = new AtomicReference<>(currentSlot);
+      this.currentIndex = new AtomicReference<>(UInt64.ZERO);
+      this.maxSlot = maxSlot;
+    }
+
+    SafeFuture<Void> sendBlobSidecar(final BlobSidecar blobSidecar) {
+      return callback.respond(blobSidecar).thenRun(sentBlobSidecars::incrementAndGet);
+    }
+
+    SafeFuture<Optional<BlobSidecar>> loadNextBlobSidecar() {
+      // currentBlock is used to avoid querying the combinedChainDataClient for the same slot again
+      if (maybeCurrentBlock.isEmpty()) {
+        return combinedChainDataClient
+            .getBlockAtSlotExact(currentSlot.get(), headBlockRoot)
+            .thenCompose(
+                block -> {
+                  maybeCurrentBlock = block;
+                  return retrieveBlobSidecar();
+                });
+      } else {
+        return retrieveBlobSidecar();
+      }
+    }
+
+    boolean isComplete() {
+      return (currentSlot.get().equals(maxSlot)
+              && currentIndex.get().equals(maxBlobsPerBlock.minus(UInt64.ONE)))
+          || maxRequestSize.isLessThanOrEqualTo(sentBlobSidecars.get());
+    }
+
+    void incrementSlotAndIndex() {
+      if (currentIndex.get().equals(maxBlobsPerBlock.minus(UInt64.ONE))) {
+        currentIndex.set(UInt64.ZERO);
+        currentSlot.updateAndGet(UInt64::increment);
+      } else {
+        currentIndex.updateAndGet(UInt64::increment);
+      }
+    }
+
+    @NotNull
+    private SafeFuture<Optional<BlobSidecar>> retrieveBlobSidecar() {
+      SafeFuture<Optional<BlobSidecar>> blobSidecar =
+          maybeCurrentBlock
+              .map(SignedBeaconBlock::getRoot)
+              .map(
+                  blockRoot ->
+                      combinedChainDataClient.getBlobSidecarBySlotIndexAndBlockRoot(
```

Comment:
Are we getting only confirmed blobs here? @tbenr ?
I don't understand it completely yet

Reply:
FYI this is not yet implemented...

Reply:
I think we don't have to pass the slot since we have the root from the block so can just use getBlobSidecarByBlockRootAndIndex(). I am using the same method in the other RPC method. These could change anyways when we implement the storage so better to have one for now I think.

Reply:
@zilm13 yes the apis must filter out unconfirmed Sidecars (even if currently we don't do that)

Reply:
Done